text/0000-embrace-extend-extinguish.md

```diff
@@ -0,0 +1,96 @@
+- Feature Name: embrace-extend-extinguish
+- Start Date: 2015-02-13
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Extend the Extend trait to take IntoIterator, and make all collections
+`impl<'a, T: Clone> Extend<&'a T>`. This enables both `vec.extend(&[1, 2, 3])`, and
+`vec.extend(&hash_set)`. This provides a more expressive replacement for
+`Vec::push_all` with literally no ergonomic loss, while leveraging established APIs.
+
+# Motivation
+
+Vec::push_all is kinda random and specific. Partially motivated by performance concerns,
+but largely just "nice" to not have to do something like
+`vec.extend([1, 2, 3].iter().cloned())`. The performance argument falls flat
+(we *must* make iterators fast, and trusted_len should get us there). The ergonomics
+argument is salient, though. Working with Plain Old Data types in Rust is super annoying
+because generic APIs and semantics are tailored for non-Copy types.
+
+Even with Extend upgraded to take IntoIterator, that won't work with &[Copy],
+because a slice can't be moved out of. Collections would have to take `IntoIterator<&T>`,
+and clone out of the reference. So, do exactly that.
+
+As a bonus, this is more expressive than `push_all`, because you can feed in any
+collection by-reference to clone the data out of it.
+
+# Detailed design
+
+Here's a quick hack to get this working today:
+
+```
+/// A type growable from an `Iterator` implementation
+pub trait Extend<T> {
+    fn extend<It: Iterator<Item=T>, I: IntoIterator<IntoIter=It>>
```

Comment:
I'd be fine just dropping this definition entirely from the RFC, I think the definition below should [work shortly](https://github.com/rust-lang/rust/pull/22313) as well.